Mono.Debugging.Soft/SoftDebuggerSession.cs

```diff
@@ -737,19 +739,84 @@ protected override ProcessInfo[] OnGetProcesses ()
 		internal PortablePdbData GetPdbData (AssemblyMirror asm)
 		{
 			string assemblyFileName;
+
 			if (!assemblyPathMap.TryGetValue (asm.GetName ().FullName, out assemblyFileName))
 				assemblyFileName = asm.Location;
 			var pdbFileName = Path.ChangeExtension (assemblyFileName, ".pdb");
-			if (!PortablePdbData.IsPortablePdb (pdbFileName))
-				return null;
-			return new PortablePdbData (pdbFileName);
+			if (PortablePdbData.IsPortablePdb (pdbFileName))
+				return new PortablePdbData (pdbFileName);
+			// Attempt to fetch pdb from the debuggee over the wire
+			var pdbBlob = asm.GetPdbBlob ();
+			return pdbBlob != null ? new PortablePdbData (pdbBlob) : null;
 		}
 
 		internal PortablePdbData GetPdbData (MethodMirror method)
 		{
 			return GetPdbData (method.DeclaringType.Assembly);
 		}
 
+		Dictionary<AssemblyMirror, SourceLinkMap []> SourceLinkCache = new Dictionary<AssemblyMirror, SourceLinkMap []> ();
+		SourceLinkMap [] GetSourceLinkMaps (MethodMirror method)
+		{
+			var asm = method.DeclaringType.Assembly;
+			SourceLinkMap [] maps = Array.Empty<SourceLinkMap>();
+
+			if (SourceLinkCache.TryGetValue (asm, out maps)) {
+				return maps;
+			}
+
+			string jsonString = null;
+
+			if (asm.VirtualMachine.Version.AtLeast (2, 48)) {
+				jsonString = asm.ManifestModule.SourceLink;
+			} else {
+				// Read the pdb ourselves
+				jsonString = GetPdbData (asm)?.GetSourceLinkBlob ();
+			}
+
+			if (!string.IsNullOrWhiteSpace(jsonString)) {
+				try {
+					var jsonSourceLink = JsonConvert.DeserializeObject<JsonSourceLink> (jsonString);
+
+					if (jsonSourceLink != null && jsonSourceLink.Maps != null && jsonSourceLink.Maps.Any ())
+						maps = jsonSourceLink.Maps.Select (kv => new SourceLinkMap (kv.Key, kv.Value)).ToArray ();
+				} catch (JsonException ex) {
+					DebuggerLoggingService.LogError ("Error reading source link", ex);
+				}
+			}
+			
+			SourceLinkCache.Add (asm, maps);
```

Comment:
What's the threading story here? If 2 threads start computing the cache at once when it's empty, they can trigger an exception here:
```
t1 - no cache
t2 - no cache
t1 - add map
t2 - add map -> exception
```

Reply:
We either need a lock on the dictionary or we need to make this a:

```suggestion
			SourceLinkCache [asm] = maps;
```

Reply:
I was already ensuring that maps could not be null. The Indexer suggestion is a good one though

Reply:
I guess I'll change it anyway, maybe it's clearer your way